### ISSUE TITLE
LSTM implementation - sgd not converging

### ISSUE THREAD

**Issue by kayoyin** [mlpack/mlpack#1605]
Dear all,
I am trying to implement a LSTM model for music composition, but I have to admit I'm struggling to see how everything is supposed to be implemented. 
I'm facing matrice index issues at
```
Breakpoint 3, mlpack::ann::NegativeLogLikelihood<arma::Mat<double>, arma::Mat<double> >::Forward<arma::Mat<double>, arma::Mat<double> > (
    this=0x7fffffffdae0, input=..., target=...) at /usr/local/include/mlpack/methods/ann/layer/negative_log_likelihood_impl.hpp:39
39	    output -= input(currentTarget, i);
```
and I have `currentTarget = 21` while `input` only has 1 row.
I see that `currentTarget` is defined during the implementation of negative log likelihood as `size_t currentTarget = target(i) - 1` but I'm not sure what this `target` is.

While debugging, I get that `target` is 
```
(arma::Mat<double> &&) @0x7fffffffc440: {<arma::Base<double, arma::Mat<double> >> = {<arma::Base_inv_yes<arma::Mat<double> >> = {<No data fields>}, <arma::Base_eval_Mat<double, arma::Mat<double> >> = {<No data fields>}, <arma::Base_trans_default<arma::Mat<double> >> = {<No data fields>}, <No data fields>}, n_rows = 1, n_cols = 5, n_elem = 5, vec_state = 0, 
  mem_state = 2, mem = 0x555555cbb720, mem_local = {1.5810100666919889e-322, 6.9533558070907438e-310, 
    6.9533482632581281e-310, 0, 0, 6.9533480903420196e-310, 0, 1.5810100666919889e-322, 1.0690394818454556e-318, 
    6.9533558070907438e-310, 6.9533558074990397e-310, 0, 0, 6.9533480914720465e-310, 4.6355707957136222e-310, 
    6.9533558071085302e-310}, static is_col = false, static is_row = false}
```

**Comment by zoq** [comment]
Happy to help if you could provide some more information like a simple model with example data to reproduce the issue.

**Comment by kayoyin** [comment]
Thank you so much!
You can find the code for our model here: https://github.com/RayDiab/Prodigy/blob/master/src/main.cpp
There's a sample data under /utils called `training.csv`, and a CMake file in the same repo
Thank you for your help!

**Comment by kayoyin** [comment]
To update you, I changed the loss function to `CrossEntropyError` and the training phase has no more issues during runtime. Now, I'm faced with this error though: `SGD: converged to -nan; terminating with failure.  Try a smaller step size?` so the accuracy of my training is always 0%.

Here's how I built the model
```
RNN<CrossEntropyError<> > model(rho);
    model.Add<IdentityLayer<> > ();
    model.Add<LSTM <> > (trainX.n_rows, 4, rho);
    model.Add<LSTM <> > (4,4);
    model.Add<LSTM <> > (4,4);
    model.Add<LSTM <> > (4,4);
    model.Add<Linear <> > (4, 1);
```

**Comment by rcurtin** [comment]
What step size did you use fpr SGD?  Have ypu tried a smaller step size?

**Comment by zoq** [comment]
In this case it might be useful to use: the ``MeanSquaredError`` loss function.

**Comment by kayoyin** [comment]
I tried with `MeanSquaredError` and I no longer have the SGD not converging error!
However, I have 0 accuracy in training after 10 training cycles, I'm not sure if it's from an error in our implementation or if we simply need to train more.
Here's the output of `model.Predict()`:
```
[cube slice 0]
  -1.3131  -1.3125  -1.3144  -1.3159  -1.3163  -1.3163  -1.3161  -1.3159  -1.3158  -1.3159  -1.3165  -1.3165  -1.3163  -1.3168  -1.3168  -1.3162

[cube slice 1]
  -1.3396  -1.3387  -1.3381  -1.3377  -1.3371  -1.3367  -1.3365  -1.3365  -1.3363  -1.3376  -1.3381  -1.3381  -1.3372  -1.3374  -1.3374  -1.3367

[cube slice 2]
  -1.3559  -1.3562  -1.3556  -1.3550  -1.3542  -1.3540  -1.3538  -1.3540  -1.3537  -1.3549  -1.3553  -1.3553  -1.3546  -1.3545  -1.3546  -1.3541

[cube slice 3]
  -1.3635  -1.3642  -1.3637  -1.3633  -1.3625  -1.3624  -1.3622  -1.3627  -1.3622  -1.3631  -1.3634  -1.3635  -1.3631  -1.3628  -1.3628  -1.3626

[cube slice 4]
  -1.3659  -1.3667  -1.3663  -1.3661  -1.3653  -1.3653  -1.3652  -1.3658  -1.3652  -1.3659  -1.3661  -1.3663  -1.3662  -1.3656  -1.3657  -1.3656

[cube slice 5]
  -1.3660  -1.3668  -1.3665  -1.3665  -1.3658  -1.3659  -1.3657  -1.3665  -1.3657  -1.3663  -1.3665  -1.3666  -1.3667  -1.3661  -1.3661  -1.3662

[cube slice 6]
  -1.3655  -1.3661  -1.3659  -1.3660  -1.3654  -1.3656  -1.3654  -1.3662  -1.3654  -1.3658  -1.3660  -1.3661  -1.3663  -1.3657  -1.3657  -1.3659

[cube slice 7]
  -1.3648  -1.3653  -1.3652  -1.3654  -1.3649  -1.3651  -1.3650  -1.3656  -1.3649  -1.3652  -1.3653  -1.3655  -1.3657  -1.3651  -1.3652  -1.3653
```

While the actual labels are 
```
[cube slice 0]
   22.0000   16.0000   19.0000   11.0000   22.0000   14.0000   13.0000   11.0000    8.0000   16.0000   15.0000   12.0000    9.0000   16.0000   19.0000   11.0000
```
And to calculate accuracy I'm counting the percentage of integers (notes) that match between the last slice of the predicted output and the first slice of the actual labels.
 
I was also wondering if my training data is in a suitable format for training the LSTM model?

**Comment by rcurtin** [comment]
Four layers of LSTMs is quite a lot:

```
RNN<MeanSquaredError<> > model(rho);
model.Add<Linear <> > (trainX.n_rows, rho);
model.Add<LSTM <> > (rho,4);
model.Add<LSTM <> > (4,4);
model.Add<LSTM <> > (4,4);
model.Add<LSTM <> > (4,4);
model.Add<SigmoidLayer <> >();
model.Add<Linear <> > (4, 1);
```

I'd suggest a handful of changes:

1. Start with only one LSTM layer.  Four layers deep will need a huge amount of data to train and that could be difficult.  Maybe one LSTM layer won't perform amazingly but it will at least provide a starting point.
2. It looks like you're only using 4 LSTM units in a single layer, which is probably a little low.  You might try more like 50.
3. Sometimes it can help to add a Linear layer after the LSTMs but before the sigmoid.

I hope this helps, but I doubt that there's a bug in the mlpack code based on what you've written, and probably you'll see better performance by tuning the network structure and how you are learning. :+1:

**Comment by rcurtin** [comment]
> I ran the exact same code without the new changes on the second VM but the SGD doesn't converge, while on my first VM it converged.

Probably the network is using a different random initialization.  You could try setting the random seed differently, or again---make the step size of SGD a little bit smaller.  Or, you could also try a different optimizer entirely.

> Also, for the format of my data, do I need to format the memory aspect of the RNN?

So, the way you have it now can work, but also you could consider one-hot encoding the notes.  i.e., if your set of possible notes was (just to take a short example) [C C# D Eb E F F# G] then each input at each timestep could be, e.g., [1 0 0 0 0 0 0 0], [0 0 0 1 0 0 0 0], [0 0 0 0 0 0 0 1].  Honestly I'd say try both things and see what works best.  My experience with deep learning has been that experimenting with all possible ideas is the best way to find the best solution. :+1:

**Comment by kayoyin** [comment]
Thank you! I got the training working for now :) It's still very slow so I'll have to experiment with different parameters etc.

Now, how do you save a model that has been trained? Is there a way to save the weights, and load the weights into a model for later use?

Thank you so much for all your help so far!

**Comment by rcurtin** [comment]
Glad you got it working.  To save a model, you can see the documentation: http://mlpack.org/docs/mlpack-git/doxygen/formatdoc.html
Specifically, the "loading and saving models in C++" part is what I think you are looking for.

**Comment by zoq** [comment]
The Neural Network tutorial has also a section about Saving/Loading models: https://www.mlpack.org/docs/mlpack-3.0.4/doxygen/anntutorial.html#model_saving_loading_anntut

**Comment by kayoyin** [comment]
Sorry I should have checked out and found the documentation but thank you so much!
I have another question sorry, I was thinking of representing the target as a vector of 0s except at the index of the value to predict we have 1. Does mlpack have a loss function, something like categorical cross entropy, to support that format?

**Comment by rcurtin** [comment]
Right, for this you can use `NegativeLogLikelihood`.  But there is no special need to change your labels---you can represent them just as a vector e.g. `[5, 3, 2, 2, 1]` or similar.  Note that the labels should start from 1, not 0.

**Comment by kayoyin** [comment]
As the `Train()` method of the RNN model takes in two armadillo cubes as inputs, how would the labels `[5, 3, 2, 2, 1]` be reformatted into a cube? What should each slice contain here?

**Comment by rcurtin** [comment]
Each slice corresponds to a sequence, and in each slice a column corresponds to a single time step.  So if a sequence has 10 time steps, and you have 100 sequences total, then your cube would have 100 slices, 10 columns, and 1 row.  That one-element row would contain the label for each time step.

The documentation for all this could definitely be improved, so thanks for sticking with it.  I hope to improve it at some point. :)

**Comment by kayoyin** [comment]
Thank you! So I tried with `NegativeLogLikelihood` with the labels containing the actual values as you suggested, but now the training doesn't seem to be working as my model is predicting the exact same thing at each training cycle, do you have an idea of what I might have done wrong?

**Comment by rcurtin** [comment]
It could be a step size issue or another optimizer configuration issue---it's really hard to know.  It sounds like you have the data configured right, so I think your best bet here might be experimenting with different optimizer (and network) configurations and seeing what works and what doesn't.  :+1:
